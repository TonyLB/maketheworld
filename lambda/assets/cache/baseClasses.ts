import { ComponentRenderItem } from "@tonylb/mtw-wml/dist/normalize"

export type EphemeraCondition = {
    dependencies: string[];
    if: string;
}

export type EphemeraFeatureAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
}

export type EphemeraFeature = {
    EphemeraId: string;
    key: string;
    tag: 'Feature';
    appearances: EphemeraFeatureAppearance[];
}

export type EphemeraExit = {
    name: string;
    to: string;
}

export type EphemeraRoomAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
    exits: EphemeraExit[];
}

export type EphemeraRoom = {
    EphemeraId: string;
    key: string;
    tag: 'Room';
    appearances: EphemeraRoomAppearance[];
}

export type EphemeraMapRoom = {
    EphemeraId: string;
    x: number;
    y: number;
}

export type EphemeraMapAppearance = {
    conditions: EphemeraCondition[];
    fileURL: string;
    name: string;
    rooms: Record<string, EphemeraMapRoom>;
}

export type EphemeraMap = {
    EphemeraId: string;
    key: string;
    tag: 'Map';
    appearance: EphemeraMapAppearance;
}

//
// TODO (ISS1385): Add EphemeraAction to base classes
//

export type EphemeraItem = EphemeraFeature | EphemeraRoom | EphemeraMap
