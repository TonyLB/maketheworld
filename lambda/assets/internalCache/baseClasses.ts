export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
