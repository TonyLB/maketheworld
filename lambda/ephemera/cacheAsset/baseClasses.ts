import { AssetWorkspaceAddress } from "@tonylb/mtw-asset-workspace/dist"
import {
    EphemeraActionId,
    EphemeraCharacterId,
    EphemeraComputedId,
    EphemeraFeatureId,
    EphemeraMapId,
    EphemeraRoomId,
    EphemeraVariableId
} from "@tonylb/mtw-interfaces/dist/ephemera";
import { splitType } from "@tonylb/mtw-utilities/dist/types";
import { ComponentRenderItem, NormalCharacterPronouns } from "@tonylb/mtw-wml/dist/normalize/baseClasses"

export type EphemeraItemDependency = {
    key: string;
    EphemeraId: string;
}

export type EphemeraCondition = {
    dependencies: EphemeraItemDependency[];
    if: string;
}

export type EphemeraFeatureAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
}

export type EphemeraFeature = {
    EphemeraId: EphemeraFeatureId;
    key: string;
    appearances: EphemeraFeatureAppearance[];
}

export type EphemeraExit = {
    name: string;
    to: string;
}

export type EphemeraRoomAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
    exits: EphemeraExit[];
}

export type EphemeraRoom = {
    EphemeraId: EphemeraRoomId;
    key: string;
    appearances: EphemeraRoomAppearance[];
}

export type EphemeraMapRoom = {
    EphemeraId: string;
    x: number;
    y: number;
}

export type EphemeraMapAppearance = {
    conditions: EphemeraCondition[];
    fileURL: string;
    name: string;
    rooms: Record<string, EphemeraMapRoom>;
}

export type EphemeraMap = {
    EphemeraId: EphemeraMapId;
    key: string;
    appearances: EphemeraMapAppearance[];
}

export type EphemeraCharacter = {
    EphemeraId: EphemeraCharacterId;
    key: string;
    address: AssetWorkspaceAddress;
    Name: string;
    Pronouns: NormalCharacterPronouns;
    FirstImpression: string;
    OneCoolThing?: string;
    Outfit?: string;
    Color: 'blue' | 'purple' | 'green' | 'pink';
    fileURL?: string;
    Connected: boolean;
    ConnectionIds: string[];
    RoomId: string;
}

export type EphemeraAction = {
    EphemeraId: EphemeraActionId;
    key: string;
    src: string;
}

export type EphemeraVariable = {
    EphemeraId: EphemeraVariableId;
    key: string;
    default: string;
}

export type EphemeraComputed = {
    EphemeraId: EphemeraComputedId;
    key: string;
    src: string;
    dependencies: EphemeraItemDependency[];
}

export type EphemeraItem = EphemeraFeature | EphemeraRoom | EphemeraMap | EphemeraCharacter | EphemeraAction | EphemeraVariable | EphemeraComputed

type LegalEphemeraTag = 'Asset' | (EphemeraItem['EphemeraId'] extends `${infer T}#${string}` ? Capitalize<Lowercase<T>> : never)

const isLegalEphemeraTag = (tag: string): tag is LegalEphemeraTag => (['Asset', 'Room', 'Map', 'Character', 'Action', 'Variable', 'Computed'].includes(tag))

export const tagFromEphemeraWrappedId = (EphemeraId: string): LegalEphemeraTag => {
    const [upperTag] = splitType(EphemeraId)
    const tag = `${upperTag[0].toUpperCase()}${upperTag.slice(1).toLowerCase()}`
    if (isLegalEphemeraTag(tag)) {
        return tag
    }
    else {
        throw new Error(`Invalid dependency tag: ${tag}`)
    }
}

export type EphemeraDependencyImport = {
    key: string;
    asset: string;
}

export type EphemeraDependencies = {
    room?: string[];
    map?: string[];
    mapCache?: string[];
    computed?: string[];
    imported?: EphemeraDependencyImport[];
}

export type EphemeraStateComputed = {
    key: string;
    computed: true;
    src: string;
    value?: any;
}

export type EphemeraStateVariable = {
    key: string;
    computed?: false;
    imported?: false;
    value: any;
}

export type EphemeraStateImport = {
    key: string;
    computed?: false;
    imported: true;
    value: any;
    asset: string;
}

type EphemeraStateItem = EphemeraStateComputed | EphemeraStateVariable | EphemeraStateImport

export type EphemeraState = {
    [key: string]: EphemeraStateItem
}

export const isEphemeraStateVariable = (item: EphemeraStateItem): item is EphemeraStateVariable => (!(item.computed || item.imported))
export const isEphemeraStateComputed = (item: EphemeraStateItem): item is EphemeraStateComputed => (item.computed || false)
export const isEphemeraStateImport = (item: EphemeraStateItem): item is EphemeraStateImport => (!(item.computed || false) && (item.imported || false))

export type EphemeraImportStateItem = {
    state: EphemeraState;
    dependencies: Record<string, EphemeraDependencies>;
}

export type EphemeraImportState = {
    [key: string]: EphemeraImportStateItem;
}
export type EphemeraPushArgs = {
    EphemeraId: string;
    State: EphemeraState;
    Dependencies?: EphemeraDependencies;
    Actions?: any;
    importTree?: any;
    scopeMap?: any;
}