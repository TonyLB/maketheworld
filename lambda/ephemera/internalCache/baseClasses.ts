import { EphemeraCharacterId, EphemeraRoomId, LegalCharacterColor } from "@tonylb/mtw-interfaces/dist/baseClasses";

export class Deferred <T>{
    invalidationCounter: number;
    promise: Promise<T>;
    _resolve: (value: T) => void = () => {}
    _reject: () => void = () => {}
    isFulfilled: boolean = false
    isRejected: boolean = false
    constructor() {
        this.invalidationCounter = 0
        this.promise = new Promise((resolve, reject)=> {
            this._reject = reject
            this._resolve = resolve
        })
    }

    invalidate() {
        this.invalidationCounter += 1
        if (this.isFulfilled || this.isRejected) {
            this.promise = new Promise((resolve, reject)=> {
                this._reject = reject
                this._resolve = resolve
            })
            this.isFulfilled = false
            this.isRejected = false
        }
    }

    resolve (invalidationCounter: number, value: T): boolean {
        if (invalidationCounter < this.invalidationCounter) {
            return false
        }
        this._resolve(value)
        this.isFulfilled = true
        return true
    }

    reject (invalidationCounter: number): void {
        if (invalidationCounter >= this.invalidationCounter) {
            this._reject()
            this.isRejected = true
        }
    }
}

export type RoomCharacterListItem = {
    EphemeraId: EphemeraCharacterId;
    Color?: LegalCharacterColor;
    ConnectionIds: string[];
    fileURL?: string;
    Name: string;
}

export type DependencyEdge = {
    EphemeraId: string;
    key?: string;
    assets: string[];
}

export type DependencyNode = {
    EphemeraId: string;
    completeness: 'Partial' | 'Complete';
    connections: DependencyEdge[]
}

export type LegalDependencyTag = 'Asset' | 'Variable' | 'Computed' | 'Room' | 'Feature' | 'Map'
export const isLegalDependencyTag = (tag: string): tag is LegalDependencyTag => (['Asset', 'Variable', 'Computed', 'Room', 'Feature', 'Map'].includes(tag))

export type DependencyGraphPut = {
    EphemeraId: string;
    putItem: DependencyEdge;
}

export type DependencyGraphDelete = {
    EphemeraId: string;
    deleteItem: DependencyEdge;
}

export type DependencyGraphAction = DependencyGraphPut | DependencyGraphDelete

export const isDependencyGraphPut = (action: DependencyGraphAction): action is DependencyGraphPut => ('putItem' in action)
export const isDependencyGraphDelete = (action: DependencyGraphAction): action is DependencyGraphDelete => ('deleteItem' in action)

export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
