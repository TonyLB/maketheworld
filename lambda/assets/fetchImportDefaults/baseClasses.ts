import { EphemeraAssetId } from "@tonylb/mtw-interfaces/ts/baseClasses"
import { JSONFileCache } from "../internalCache/jsonFile"
import { Graph } from "@tonylb/mtw-utilities/dist/graphStorage/utils/graph"
import { AssetWorkspaceAddress } from "@tonylb/mtw-asset-workspace"
import ReadOnlyAssetWorkspace from "@tonylb/mtw-asset-workspace/dist/readOnly"

export class InheritanceGraph extends Graph<EphemeraAssetId, { key: EphemeraAssetId; address: AssetWorkspaceAddress }, {}> {}

export class FetchImportsJSONHelper {
    _inheritanceGraph: InheritanceGraph
    constructor(inheritanceGraph: InheritanceGraph) {
        this._inheritanceGraph = inheritanceGraph
    }

    async get(assetId: EphemeraAssetId): Promise<JSONFileCache> {
        const node = this._inheritanceGraph.nodes[assetId]
        if (node) {
            const assetWorkspace = new ReadOnlyAssetWorkspace(node.address)
            await assetWorkspace.loadJSON()
            return {
                normal: assetWorkspace.normal || {},
                namespaceIdToDB: assetWorkspace.namespaceIdToDB || {}
            }    
        }
        else {
            return {
                normal: {},
                namespaceIdToDB: {}
            }
        }
    }
}

