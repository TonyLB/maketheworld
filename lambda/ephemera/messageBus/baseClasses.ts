import { AttributeValue } from "@aws-sdk/client-dynamodb"
import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace/dist'
import { EventBridgeUpdatePlayerCharacter, EventBridgeUpdatePlayerAsset } from '@tonylb/mtw-interfaces/dist/eventBridge'
import { TaggedMessageContent, LegalCharacterColor, FeatureDescription, RoomDescription, CharacterDescription } from "@tonylb/mtw-interfaces/dist/messages"
import { DependencyEdge, DependencyGraphAction, DependencyNode, RoomCharacterListItem } from "../internalCache/baseClasses"
import { EphemeraCharacterId, EphemeraExit, EphemeraFeatureId, EphemeraMapId, EphemeraRoomId } from "../cacheAsset/baseClasses"

export type PublishTargetRoom = {
    roomId: string;
}

export type PublishTargetCharacter = {
    characterId: string;
}

export type PublishTargetExcludeCharacter = {
    excludeCharacterId: string;
}

export type PublishTarget = PublishTargetRoom | PublishTargetCharacter | PublishTargetExcludeCharacter

export const isPublishTargetRoom = (prop: PublishTarget): prop is PublishTargetRoom => (('roomId' in prop) && Boolean(prop.roomId))
export const isPublishTargetCharacter = (prop: PublishTarget): prop is PublishTargetCharacter => (('characterId' in prop) && Boolean(prop.characterId))
export const isPublishTargetExcludeCharacter = (prop: PublishTarget): prop is PublishTargetExcludeCharacter => (('excludeCharacterId' in prop) && Boolean(prop.excludeCharacterId))

export type PublishMessageBase = {
    type: 'PublishMessage';
    targets: PublishTarget[];
}

export type PublishWorldMessage = PublishMessageBase & {
    displayProtocol: 'WorldMessage';
    message: TaggedMessageContent[];
}

type MessageCharacterInfo = {
    characterId: string;
    name: string;
    color: LegalCharacterColor;
}

export type PublishSpeechMessage = PublishMessageBase & MessageCharacterInfo & {
    displayProtocol: 'SayMessage';
    message: TaggedMessageContent[];
}

export type PublishNarrateMessage = PublishMessageBase & MessageCharacterInfo & {
    displayProtocol: 'NarrateMessage';
    message: TaggedMessageContent[];
}

export type PublishOutOfCharacterMessage = {
    displayProtocol: 'OOCMessage';
    message: TaggedMessageContent[];
} & PublishMessageBase & MessageCharacterInfo

export type PublishRoomUpdateMessage = {
    displayProtocol: 'RoomUpdate';
    RoomId: string;
    Characters: (Omit<RoomCharacterListItem, 'EphemeraId' | 'ConnectionIds'> & { CharacterId: string })[];
} & PublishMessageBase

export type PublishFeatureDescriptionMessage = Omit<FeatureDescription, 'DisplayProtocol' | 'MessageId' | 'CreatedTime' | 'Target'> & {
    displayProtocol: 'FeatureDescription';
} & PublishMessageBase

export type PublishRoomDescriptionMessage = Omit<RoomDescription, 'DisplayProtocol' | 'MessageId' | 'CreatedTime' | 'Target'> & {
    displayProtocol: 'RoomDescription';
} & PublishMessageBase

export type PublishCharacterDescriptionMessage = Omit<CharacterDescription, 'DisplayProtocol' | 'MessageId' | 'CreatedTime' | 'Target'> & {
    displayProtocol: 'CharacterDescription';
} & PublishMessageBase

export type PublishMessage = PublishWorldMessage |
    PublishSpeechMessage |
    PublishNarrateMessage |
    PublishOutOfCharacterMessage |
    PublishRoomUpdateMessage |
    PublishFeatureDescriptionMessage |
    PublishRoomDescriptionMessage |
    PublishCharacterDescriptionMessage

export type ReturnValueMessage = {
    type: 'ReturnValue';
    body: Record<string, any>;
}

export type DisconnectMessage = {
    type: 'Disconnect';
    connectionId: string;
}

export type ConnectMessage = {
    type: 'Connect';
    userName: string;
}

export type WhoAmIMessage = {
    type: 'WhoAmI'
}

export type SyncRequest = {
    type: 'Sync';
    targetId: string;
    LastEvaluatedKey?: Record<string, AttributeValue>;
    startingAt?: number;
    limit?: number;
    loopCount?: number;
}

export type SyncResponse = {
    type: 'SyncResponse',
    messages: any[];
    lastSync?: number;
}

export type RegisterCharacterMessage = {
    type: 'RegisterCharacter';
    characterId: string;
}

export type EphemeraUpdateCharacterEntry = {
    type: 'CharacterInPlay';
    CharacterId: string;
    Connected: boolean;
    RoomId: string;
    Name: string;
    fileURL: string;
    Color: string;
}

export type EphemeraUpdateMapEntry = {
    type: 'MapUpdate';
    targets: PublishTarget[];
    MapId: string;
    Name: string;
    fileURL: string;
    rooms: {
        roomId: string;
        name: string;
        x: number;
        y: number;
        exits: EphemeraExit[];
    }[]
}

export type EphemeraUpdateEntry = EphemeraUpdateCharacterEntry | EphemeraUpdateMapEntry;

export type EphemeraUpdateMessage = {
    type: 'EphemeraUpdate';
    global: boolean;
    updates: EphemeraUpdateEntry[];
}

export type FetchPlayerEphemeraMessage = {
    type: 'FetchPlayerEphemera';
}

export type ImportDefaultsMessage = {
    type: 'ImportDefaults';
    components: Record<string, any>;
    aggregateExits: any[];
}

export type FetchImportDefaultsMessage = {
    type: 'FetchImportDefaults';
    importsByAssetId: Record<string, any>;
    assetId: string;
}

export type PerceptionMessage = {
    type: 'Perception';
    characterId: string;
    ephemeraId: EphemeraRoomId | EphemeraFeatureId | EphemeraCharacterId | EphemeraMapId;
}

export type MoveCharacterMessage = {
    type: 'MoveCharacter';
    characterId: string;
    roomId: string;
    leaveMessage?: string;
}

export type DecacheAssetMessage = {
    type: 'DecacheAsset';
    assetId: string;
}

type CacheAssetOptions = {
    check?: boolean;
    recursive?: boolean;
    forceCache?: boolean;
}

export type CacheAssetMessage = {
    type: 'CacheAsset';
    address: AssetWorkspaceAddress;
    options: CacheAssetOptions;
}

export type PlayerUpdateMessage = {
    type: 'PlayerUpdate';
    player: string;
    Characters: EventBridgeUpdatePlayerCharacter[];
    Assets: EventBridgeUpdatePlayerAsset[];
}

export type RoomUpdateMessage = {
    type: 'RoomUpdate';
    roomId: string;
}

export type LegalDependencyTag = 'Asset' | 'Variable' | 'Computed' | 'Room' | 'Feature' | 'Map'
export const isLegalDependencyTag = (tag: string): tag is LegalDependencyTag => (['Asset', 'Variable', 'Computed', 'Room', 'Feature', 'Map'].includes(tag))

export type DescentUpdateMessage = {
    type: 'DescentUpdate';
} & DependencyGraphAction

export type AncestryUpdateMessage = {
    type: 'AncestryUpdate';
} & DependencyGraphAction

export type DependencyCascadeMessage = {
    type: 'DependencyCascade';
    targetId: string;
}

export type ExecuteActionMessage = {
    type: 'ExecuteAction';
    actionId: string;
    characterId: string;
}

export type MessageType = PublishMessage |
    ReturnValueMessage |
    DisconnectMessage |
    ConnectMessage |
    WhoAmIMessage |
    SyncRequest |
    SyncResponse |
    RegisterCharacterMessage |
    EphemeraUpdateMessage |
    FetchPlayerEphemeraMessage |
    ImportDefaultsMessage |
    FetchImportDefaultsMessage |
    PerceptionMessage |
    MoveCharacterMessage |
    DecacheAssetMessage |
    CacheAssetMessage |
    PlayerUpdateMessage |
    RoomUpdateMessage |
    DescentUpdateMessage |
    AncestryUpdateMessage |
    DependencyCascadeMessage |
    ExecuteActionMessage

export const isPublishMessage = (prop: MessageType): prop is PublishMessage => (prop.type === 'PublishMessage')
export const isWorldMessage = (prop: PublishMessage): prop is PublishWorldMessage => (prop.displayProtocol === 'WorldMessage')
export const isCharacterMessage = (prop: PublishMessage): prop is (PublishSpeechMessage | PublishNarrateMessage | PublishOutOfCharacterMessage) => (['SayMessage', 'NarrateMessage', 'OOCMessage'].includes(prop.displayProtocol))
export const isRoomUpdatePublishMessage = (prop: PublishMessage): prop is PublishRoomUpdateMessage => (prop.displayProtocol === 'RoomUpdate')
export const isRoomDescriptionPublishMessage = (prop: PublishMessage): prop is PublishRoomDescriptionMessage => (prop.displayProtocol === 'RoomDescription')
export const isFeatureDescriptionPublishMessage = (prop: PublishMessage): prop is PublishFeatureDescriptionMessage => (prop.displayProtocol === 'FeatureDescription')
export const isCharacterDescriptionPublishMessage = (prop: PublishMessage): prop is PublishCharacterDescriptionMessage => (prop.displayProtocol === 'CharacterDescription')

export const isReturnValueMessage = (prop: MessageType): prop is ReturnValueMessage => (prop.type === 'ReturnValue')
export const isDisconnectMessage = (prop: MessageType): prop is DisconnectMessage => (prop.type === 'Disconnect')
export const isConnectMessage = (prop: MessageType): prop is ConnectMessage => (prop.type === 'Connect')
export const isWhoAmIMessage = (prop: MessageType): prop is WhoAmIMessage => (prop.type === 'WhoAmI')

export const isSyncRequest = (prop: MessageType): prop is SyncRequest => (prop.type === 'Sync')
export const isSyncResponse = (prop: MessageType): prop is SyncResponse => (prop.type === 'SyncResponse')
export const isRegisterCharacterMessage = (prop: MessageType): prop is RegisterCharacterMessage => (prop.type === 'RegisterCharacter')

export const isEphemeraUpdate = (prop: MessageType): prop is EphemeraUpdateMessage => (prop.type === 'EphemeraUpdate')
export const isFetchPlayerEphemera = (prop: MessageType): prop is FetchPlayerEphemeraMessage => (prop.type === 'FetchPlayerEphemera')
export const isImportDefaults = (prop: MessageType): prop is ImportDefaultsMessage => (prop.type === 'ImportDefaults')
export const isFetchImportDefaults = (prop: MessageType): prop is FetchImportDefaultsMessage => (prop.type === 'FetchImportDefaults')

export const isPerception = (prop: MessageType): prop is PerceptionMessage => (prop.type === 'Perception')
export const isMoveCharacter = (prop: MessageType): prop is MoveCharacterMessage => (prop.type === 'MoveCharacter')

export const isDecacheAsset = (prop: MessageType): prop is DecacheAssetMessage => (prop.type === 'DecacheAsset')
export const isCacheAssetMessage = (prop: MessageType): prop is CacheAssetMessage => (prop.type === 'CacheAsset')

export const isPlayerUpdateMessage = (prop: MessageType): prop is PlayerUpdateMessage => (prop.type === 'PlayerUpdate')
export const isRoomUpdateMessage = (prop: MessageType): prop is RoomUpdateMessage => (prop.type === 'RoomUpdate')

export const isDescentUpdateMessage = (prop: MessageType): prop is DescentUpdateMessage => (prop.type === 'DescentUpdate')
export const isAncestryUpdateMessage = (prop: MessageType): prop is AncestryUpdateMessage => (prop.type === 'AncestryUpdate')
export const isDependencyCascadeMessage = (prop: MessageType): prop is DependencyCascadeMessage => (prop.type === 'DependencyCascade')
export const isExecuteActionMessage = (prop: MessageType): prop is ExecuteActionMessage => (prop.type === 'ExecuteAction')

export class MessageBus extends InternalMessageBus<MessageType> {}
