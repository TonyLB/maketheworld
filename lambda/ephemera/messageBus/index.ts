import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'

import {
    MessageBus,
    isPublishMessage,
    isEphemeraUpdate,
    isDisconnectMessage,
    isConnectMessage,
    isWhoAmIMessage,
    isSyncRequest,
    isSyncResponse,
    isRegisterCharacterMessage,
    isFetchPlayerEphemera,
    isImportDefaults,
    isFetchImportDefaults,
    isPerception,
    isMoveCharacter,
    isDecacheAsset,
    isCacheAssetMessage,
    isPlayerUpdateMessage,
    isRoomUpdateMessage,
    isDescentUpdateMessage,
    isAncestryUpdateMessage,
    isDependencyCascadeMessage,
    isExecuteActionMessage,
    isMapSubscription,
    isMapUpdateMessage
} from "./baseClasses"

import publishMessage from '../publishMessage'
import ephemeraUpdate from '../ephemeraUpdate'
import disconnectMessage from '../disconnectMessage'
import connectMessage from '../connectMessage'
import whoAmIMessage from '../whoAmI'
import { syncRequest, syncResponse } from '../syncHandler'
import registerCharacter from '../registerCharacter'
import { fetchPlayerEphemera } from '../fetchEphemera'
import perceptionMessage from '../perception'
import moveCharacter from '../moveCharacter'
import decacheAssetMessage from '../decacheMessage'
import { cacheAssetMessage } from '../cacheAsset'
import playerUpdateMessage from '../playerUpdate'
import roomUpdateMessage from '../roomUpdate'
import dependentUpdateMessage from '../dependentMessages/dependentUpdate'
import dependencyCascadeMessage from '../dependentMessages/dependencyCascade'
import executeActionMessage from '../executeAction'
import mapSubscriptionMessage from '../mapSubscription'
import mapUpdateMessage from '../mapUpdate'

export const messageBus = new MessageBus()
messageBus.subscribe({
    tag: 'PublishMessage',
    priority: 15,
    filter: isPublishMessage,
    callback: publishMessage
})
messageBus.subscribe({
    tag: 'EphemeraUpdate',
    priority: 10,
    filter: isEphemeraUpdate,
    callback: ephemeraUpdate
})
messageBus.subscribe({
    tag: 'Disconnect',
    priority: 1,
    filter: isDisconnectMessage,
    callback: disconnectMessage
})
messageBus.subscribe({
    tag: 'Connect',
    priority: 1,
    filter: isConnectMessage,
    callback: connectMessage
})
messageBus.subscribe({
    tag: 'WhoAmI',
    priority: 1,
    filter: isWhoAmIMessage,
    callback: whoAmIMessage
})
messageBus.subscribe({
    tag: 'SyncResponse',
    priority: 2,
    filter: isSyncResponse,
    callback: syncResponse
})
messageBus.subscribe({
    tag: 'SyncRequest',
    priority: 3,
    filter: isSyncRequest,
    callback: syncRequest
})
messageBus.subscribe({
    tag: 'RegisterCharacter',
    priority: 1,
    filter: isRegisterCharacterMessage,
    callback: registerCharacter
})
messageBus.subscribe({
    tag: 'FetchPlayerEphemera',
    priority: 2,
    filter: isFetchPlayerEphemera,
    callback: fetchPlayerEphemera
})
messageBus.subscribe({
    tag: 'MapSubscription',
    priority: 2,
    filter: isMapSubscription,
    callback: mapSubscriptionMessage
})
messageBus.subscribe({
    tag: 'Perception',
    priority: 10,
    filter: isPerception,
    callback: perceptionMessage
})
messageBus.subscribe({
    tag: 'MoveCharacter',
    priority: 4,
    filter: isMoveCharacter,
    callback: moveCharacter
})
messageBus.subscribe({
    tag: 'DecacheAsset',
    priority: 1,
    filter: isDecacheAsset,
    callback: decacheAssetMessage
})
messageBus.subscribe({
    tag: 'CacheAsset',
    priority: 1,
    filter: isCacheAssetMessage,
    callback: cacheAssetMessage
})
messageBus.subscribe({
    tag: 'UpdatePlayer',
    priority: 5,
    filter: isPlayerUpdateMessage,
    callback: playerUpdateMessage
})
messageBus.subscribe({
    tag: 'RoomUpdate',
    priority: 6,
    filter: isRoomUpdateMessage,
    callback: roomUpdateMessage
})
messageBus.subscribe({
    tag: 'MapUpdate',
    priority: 15,
    filter: isMapUpdateMessage,
    callback: mapUpdateMessage
})
messageBus.subscribe({
    tag: 'AncestryUpdate',
    priority: 3,
    filter: isAncestryUpdateMessage,
    callback: dependentUpdateMessage('Ancestry')
})
messageBus.subscribe({
    tag: 'DescentUpdate',
    priority: 4,
    filter: isDescentUpdateMessage,
    callback: dependentUpdateMessage('Descent')
})

messageBus.subscribe({
    tag: 'DependencyCascade',
    priority: 5,
    filter: isDependencyCascadeMessage,
    callback: dependencyCascadeMessage
})

messageBus.subscribe({
    tag: 'ExecuteAction',
    priority: 7,
    filter: isExecuteActionMessage,
    callback: executeActionMessage
})

export default messageBus