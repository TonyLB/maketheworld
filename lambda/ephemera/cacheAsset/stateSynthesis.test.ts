import { jest, describe, it, expect } from '@jest/globals'

import { BaseAppearance, ComponentAppearance, NormalForm } from '@tonylb/mtw-wml/dist/normalize/baseClasses'
import { MessageBus } from '../messageBus/baseClasses'
import StateSynthesizer from './stateSynthesis'


describe('stateSynthesis', () => {

    const messageBusMock = { send: jest.fn() } as unknown as jest.Mocked<MessageBus>

    const topLevelAppearance: BaseAppearance = {
        contextStack: [{ key: 'test', tag: 'Asset', index: 0}],
        contents: []
    }

    const testNamespaceIdToDB = {
        ABC: 'ROOM#DEF',
        power: 'VARIABLE#QRS',
        switchedOn: 'VARIABLE#TUV',
        active: 'COMPUTED#XYZ'
    }
    const testAsset: NormalForm = {
        test: {
            key: 'test',
            tag: 'Asset',
            fileName: 'test',
            appearances: [{
                contextStack: [],
                contents: [{
                    key: 'ABC',
                    tag: 'Room',
                    index: 0
                },
                {
                    key: 'If-0',
                    tag: 'If',
                    index: 0
                },
                {
                    key: 'powered',
                    tag: 'Variable',
                    index: 0
                },
                {
                    key: 'switchedOn',
                    tag: 'Variable',
                    index: 0
                },
                {
                    key: 'active',
                    tag: 'Computed',
                    index: 0
                },
                {
                    key: 'toggleSwitch',
                    tag: 'Action',
                    index: 0
                },
                {
                    key: 'Import-0',
                    tag: 'Import',
                    index: 0
                }]
            }]
        },
        ABC: {
            key: 'ABC',
            tag: 'Room',
            appearances: [{
                ...topLevelAppearance,
                name: 'Vortex',
                render: []
            },
            {
                contextStack: [{ key: 'test', tag: 'Asset', index: 0 }, { key: 'If-0', tag: 'If', index: 0 }],
                name: [{ tag: 'String', value: '(lit)' }],
                render: [{ tag: 'String', value: 'The lights are on ' }],
                contents: []
            }] as ComponentAppearance[]
        },
        power: {
            key: 'power',
            tag: 'Variable',
            default: 'false',
            appearances: [topLevelAppearance]
        },
        switchedOn: {
            key: 'switchedOn',
            tag: 'Variable',
            default: 'true',
            appearances: [topLevelAppearance]
        },
        active: {
            key: 'active',
            tag: 'Computed',
            src: 'power && switchedOn',
            dependencies: ['switchedOn', 'power'],
            appearances: [topLevelAppearance]
        },
        toggleSwitch: {
            key: 'toggleSwitch',
            tag: 'Action',
            src: 'switchedOn = !switchedOn',
            appearances: [topLevelAppearance]
        },
        ['If-0']: {
            key: 'If-0',
            tag: 'If',
            conditions: [{
                if: 'active',
                dependencies: ['active']
            }],
            appearances: [{
                ...topLevelAppearance,
                contents: [{
                    key: 'ABC',
                    tag: 'Room',
                    index: 1
                }]
            }]
        },
        ['Import-0']: {
            key: 'Import-0',
            tag: 'Import',
            from: 'BASE',
            mapping: {
                welcome: { key: 'ABC', type: 'Room' },
                power: { key: 'powered', type: 'Variable' }
            },
            appearances: [topLevelAppearance]
        }
    }

    beforeEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
    })

    it('should correctly execute sendDependencyMessages', () => {

        const testSynthesizer = new StateSynthesizer({ namespaceIdToDB: testNamespaceIdToDB, normal: testAsset } as any, messageBusMock)

        testSynthesizer.sendDependencyMessages()
        expect(messageBusMock.send.mock.calls).toMatchSnapshot()
    })
})