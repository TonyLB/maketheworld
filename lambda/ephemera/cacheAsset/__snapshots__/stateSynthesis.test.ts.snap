// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    Object {
      "putItem": Object {
        "EphemeraId": "COMPUTED#XYZ",
        "assets": Array [
          "test",
        ],
        "key": "switchedOn",
      },
      "targetId": "VARIABLE#TUV",
      "type": "DescentUpdate",
    },
  ],
  Array [
    Object {
      "putItem": Object {
        "EphemeraId": "VARIABLE#TUV",
        "assets": Array [
          "test",
        ],
        "key": "switchedOn",
      },
      "targetId": "COMPUTED#XYZ",
      "type": "AncestryUpdate",
    },
  ],
  Array [
    Object {
      "putItem": Object {
        "EphemeraId": "COMPUTED#XYZ",
        "assets": Array [
          "test",
        ],
        "key": "power",
      },
      "targetId": "VARIABLE#QRS",
      "type": "DescentUpdate",
    },
  ],
  Array [
    Object {
      "putItem": Object {
        "EphemeraId": "VARIABLE#QRS",
        "assets": Array [
          "test",
        ],
        "key": "power",
      },
      "targetId": "COMPUTED#XYZ",
      "type": "AncestryUpdate",
    },
  ],
  Array [
    Object {
      "putItem": Object {
        "EphemeraId": "ROOM#DEF",
        "assets": Array [
          "test",
        ],
        "key": "active",
      },
      "targetId": "COMPUTED#XYZ",
      "type": "DescentUpdate",
    },
  ],
  Array [
    Object {
      "putItem": Object {
        "EphemeraId": "COMPUTED#XYZ",
        "assets": Array [
          "test",
        ],
        "key": "active",
      },
      "targetId": "ROOM#DEF",
      "type": "AncestryUpdate",
    },
  ],
]
`;
